### ISSUE TITLE
Dynamips router without private config: Save doesn't work

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-server#217]
GNS3 V1.3.4, OS: OS X 10.9.5 and Linux Debian 8
Topology: 2 x 3725 without private config

Problem look exctly like issue https://github.com/GNS3/gns3-server/issues/165.

On Save only one of the configs is saved:

Log:
```
2015-06-04 18:49:47 INFO topology.py:462 Starting to save the topology (version 1.3.4)
2015-06-04 18:49:47 INFO topology.py:485 saving node: R1
2015-06-04 18:49:47 INFO topology.py:485 saving node: R2
2015-06-04 18:49:47 INFO topology.py:492 saving Link from R1 port FastEthernet0/0 to R2 port FastEthernet0/0
2015-06-04 18:49:47 INFO topology.py:499 saving server 192.168.1.11:8000
2015-06-04 18:49:47 INFO main_window.py:1339 Saving project: /home/behlers/GNS3/projects/untitled/untitled.gns3
2015-06-04 18:49:47 INFO router.py:1507 saving startup-config to configs/i2_startup-config.cfg
2015-06-04 18:49:47 ERROR route.py:150 VM error detected: <class 'gns3server.modules.dynamips.dynamips_error.DynamipsError'>
Traceback (most recent call last):
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/dynamips/nodes/router.py", line 1517, in save_configs
    with open(config_path, "wb") as f:
IsADirectoryError: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/web/route.py", line 140, in control_schema
    yield from func(request, response)
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/handlers/api/project_handler.py", line 111, in commit
    yield from project.commit()
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/project.py", line 401, in commit
    yield from module.instance().project_committed(self)
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/dynamips/__init__.py", line 213, in project_committed
    yield from vm.save_configs()
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/dynamips/nodes/router.py", line 1521, in save_configs
    raise DynamipsError("Could not save the private configuration {}: {}".format(config_path, e))
gns3server.modules.dynamips.dynamips_error.DynamipsError: Could not save the private configuration /home/behlers/GNS3/projects/untitled/project-files/dynamips/: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'
2015-06-04 18:49:47 INFO http_client.py:362 Response error: Error downloading http://192.168.1.11:8000/v1/projects/e5992eef-218a-4d24-94d2-318c167dff03/commit - server replied: Conflict
```

When saving a router config with right-click-save-config an error is shown:

Server error from 192.168.1.11:8000: R1: Could not save the private configuration /home/behlers/GNS3/projects/untitled/project-files/dynamips/: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'

Logs:
```
2015-06-04 18:52:16 INFO router.py:1507 saving startup-config to configs/i1_startup-config.cfg
2015-06-04 18:52:16 ERROR route.py:150 VM error detected: <class 'gns3server.modules.dynamips.dynamips_error.DynamipsError'>
Traceback (most recent call last):
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/dynamips/nodes/router.py", line 1517, in save_configs
    with open(config_path, "wb") as f:
IsADirectoryError: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/web/route.py", line 140, in control_schema
    yield from func(request, response)
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/handlers/api/dynamips_vm_handler.py", line 415, in save_configs
    yield from vm.save_configs()
  File "/opt/GNS3/lib/python3.4/site-packages/gns3server/modules/dynamips/nodes/router.py", line 1521, in save_configs
    raise DynamipsError("Could not save the private configuration {}: {}".format(config_path, e))
gns3server.modules.dynamips.dynamips_error.DynamipsError: Could not save the private configuration /home/behlers/GNS3/projects/untitled/project-files/dynamips/: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'
2015-06-04 18:52:16 INFO http_client.py:362 Response error: Error downloading http://192.168.1.11:8000/v1/projects/e5992eef-218a-4d24-94d2-318c167dff03/dynamips/vms/92e7a72a-65d9-4bb5-ae7d-d636aeab8abf/configs/save - server replied: Conflict
2015-06-04 18:52:16 ERROR router.py:788 error while saving R1 configs: Could not save the private configuration /home/behlers/GNS3/projects/untitled/project-files/dynamips/: [Errno 21] Is a directory: '/home/behlers/GNS3/projects/untitled/project-files/dynamips/'
```

**Comment by grossmj** [comment]
This could be related to https://community.gns3.com/thread/8553

**Comment by ehlers** [comment]
Maybe, but I created this as a new project. And only saving the configuration is a problem, everything else works. Saving the config works, when the router template has a private config.

What's strange is, that the log looks exactly like the issue https://github.com/GNS3/gns3-server/issues/165, which was solved a month ago.

This issue was reported by pwakeford222, see https://community.gns3.com/thread/8737#28684

**Comment by noplay** [comment]
Strange I just test on with a blank project  and two 3725 and I don't replicate the issue :(

The faulty code is here 
```
            if private_config_base64:
                try:
                    config = base64.b64decode(private_config_base64).decode("utf-8", errors="replace")
                    config = "!\n" + config.replace("\r", "")
                    config_path = os.path.join(module_workdir, self.private_config)
                    with open(config_path, "wb") as f:
                        log.info("saving private-config to {}".format(self.private_config))
                        f.write(config.encode("utf-8"))
                except (binascii.Error, OSError) as e:
                    raise DynamipsError("Could not save the private configuration {}: {}".format(config_path, e))
```

It seem we have a base6' config but not the file.

**Comment by noplay** [comment]
Ok I can replicate the issue.

* I push a private config
* I remove it from the topology
* I restart GNS3 and try to save 

It's probably not how users create this bug, but I can fix.

**Comment by noplay** [comment]
The bug is fix, but @grossmj  what do you think about creating a private config file if a private config is exported by Dynamips?

**Comment by ehlers** [comment]
I just tested and I confirm, that it works.

But it behaves not the way it would like it. When I leave the **initial** private config empty, I want that the router starts with an empty private config. When I myself create a private config, I would expect, that this gets saved as well. But now it doesn't get saved as a file, it's only in the nvram.

**Comment by noplay** [comment]
I think now it's better fix, the private config file is created if require

**Comment by ehlers** [comment]
That looks good !!! Thanks a lot.

It had a look to the fix for the similar bug https://github.com/GNS3/gns3-server/issues/165. That was fixed just before 1.3.3dev2. At node creation @grossmj created zero length files, if the template doesn't contain an initial config. I just re-tested 1.3.3dev2, it really worked. I tested with 1.3.3, there the bug is active again. So something between 1.3.3dev2 and 1.3.3 must disable his changes.
